### ISSUE TITLE
Feature Request: Alternate way to suppress warnings

### ISSUE THREAD

**Issue by ben-page** [petkaantonov/bluebird#1059]
1) What version of bluebird is the issue happening on? **3.3.4**
2) What platform and version? **Node v5.9.1**
3) Did this issue happen with earlier version of bluebird? **Yes**

Library authors can't suppress the warning for runaway promises that they create, because it's not the runaway Promise that needs to resolve to null, but the outer promise. Consider the following example:

```javascript
'use strict';
var Promise = require('bluebird');
Promise.config({warnings: true, longStackTraces: true});

function otherLibrary() {
    Promise.resolve()
        .catch(function() {
            //totally recover from error
        })
        .then(function () {
            return null; //try to tell Bluebird we don't need the warning
        });
}

Promise.resolve()
    .then(function () {
        otherLibrary();
    })
    .delay(500); //on Windows node doesn't flush the console stream before exiting
```

This triggers a runaway promise warning even though, `otherLibrary` is trying to indicate that it wants to suppress the warning. So consumers of the library have to be aware of this and suppress the warning at every call, if they use promises as well.

```javascript
Promise.resolve()
    .then(function () {
        otherLibrary();
        return null; //this will suppress the warning
    })
    .delay(500); //on Windows node doesn't flush the console stream before exiting
```

I realize that the warning is thrown when the runaway promise is created, not when it's resolved. So the `otherLibrary` returning null is irrelevant. But, it would be nice to have a way to suppress the warning when creating the runaway promise.

Maybe a simple call that suppresses the warning for the next promise created.
```javascript
function otherLibrary() {
    Promise.suppressRunawayWarning();
    Promise.resolve()
        .catch(function() {
            //totally recover from error
        });
}
```

Or maybe it would be better to explicitly start a new promise chain.
```javascript
function otherLibrary() {
    Promise
        .newChain()
        .resolve()
        .catch(function () {
            //totally recover from error
        });
}
```

**Comment by ben-page** [comment]
My examples are contrived to make them concise. But I'll give a specific example of why this is important.

I have a library that abstracts network communications. Internally it uses promises. But it's public interfaces uses streams and events. Non-fatal error are handled transparently. Fatal errors are caught and emitted by the error event.

The internal promises can lead to runaway promise warnings if the application uses the library also uses promises.

**Comment by spion** [comment]
As far as I know warnings are now only generated if the user attempts to continue the chain after not returning a promise:

```js
Promise.resolve()
    .then(function () {
        otherLibrary();
        // nothing is returned    
    }).then(function() {
        // attempting to continue the chain
    });
```

Maybe `.delay()` should simply be exempt from the warning?

**Comment by spion** [comment]
I understand your point, but without delay (or another chained call to `then`) bluebird doesn't (or should not) generate any warnings.

If `then` is chained to a promise created by another `then` handler that didn't return anything, thats still mostly warning-worthy, I think (regardless of whether a promise was created or not)

**Comment by ben-page** [comment]
Thank you. Now I understand what you are saying about the chained calls. My example is flawed. I'll work to come up with a better one.

**Comment by ben-page** [comment]
I'm going to close this. I'll open a new issue when I've sorted out a better example.

**Comment by spion** [comment]
Feel free to reopen it, I'm not at all convinced that I'm right.

By the way `done` *should* make it really easy to implement this, if you end the chain with `done` then clearly it doesn't matter that you are using promises as you definitely don't plan to expose them (`done` meaning "the chain ends here and I don't plan to return it")

**Comment by ben-page** [comment]
@spion, I'm ready to revisit this. I agree with everything that you've said. And in most cases it's really is enough that the warning is only displayed if the chain is continued. However, I think there are more complicate situations where something like what you suggested for `.done()` would be really helpful.

Bare with me, this isn't the most concise example. But this problem is only relevant in complex situations. So here's an example from a recent project. It's a concurrent unit testing framework. It handles both synchronous and asynchronous tests. Internally there's a lot of work to make it the API consistent regardless if the test is synchronous, uses callbacks (async or sync), or uses Promises.

The following code, produces a runaway promise warning even though all promises are actually chained internally. So the test starts and `getDb()` starts a promise chain. After, the db is ready the testing can begin. Internally, `t.notThrows()` executes (synchronously) the handler passed to it and catches synchronous errors. Since the handler returns the Promise create in `db.getItems()`, the framework appends that promise to a chain inside the test suite. The suite waits for the internal chain (of all the promises in play) to resolve and captures the errors to decide if the test passed or fails.

All of this works perfectly, but Bluebird still reports a runaway Promise. The problem is that the chain created by `getDb()` is continued inside the suite. So Bluebird reports the promises created by `db.getItems()` is a runaway promise.

```javascript
    suite.test('test', t => { //create new test, t is my assertion library
        return getDb() //returns Promise
            .then(db => {
                t.notThrows(() => {
                    return db.getItems(); //returns Promise
                });
                
                //return null; //this would disable the warning
            });
    });
```

So my objection is that is up to the the user to be aware of this caveat and return null in the test. I'd like a way on the runaway promise itself, to let Bluebird know that I'm handing it. Basically, I'd like the equivalent of `.suppressUnhandledRejections()`. So it could be called `.suppressRunawayWarning()` or `.done()` could be used for this purpose.

**Comment by ben-page** [comment]
I decided to go ahead and work on a PR for this. I did this for two reasons: to gain familiarity with Bluebird and to test if this is a viable solution to my issue. I'll post in case it's helpful for the discussion. I definitely don't mean to presume that you've agreed with my request problem or that you would pull this.

**Comment by spion** [comment]
The warning you are getting here is correct.

When will the next test begin? After `db.getItems()`? I'm afraid not - nothing is returned from `then`, therefore if something chains to this it will execute immediately after `getDb()` returns the db. 

This will cause multiple tests to run in parallel. Is that what you want?

**Comment by ErisDS** [comment]
I'd also like to ask about an additional way to suppress or configure the warnings thrown from Bluebird, but with regard to the [a promise was rejected with a non-error](https://github.com/petkaantonov/bluebird/blob/master/docs/docs/warning-explanations.md#warning-a-promise-was-rejected-with-a-non-error) warning.

I agree with, and appreciate the opinions that bluebird enforces about not rejecting promises with primitive values. However, in Ghost (and other similar API-providing software) we sometimes want to reject promises with an Array of Errors, rather than a single Error. This is specifically useful when doing validations where we want to validate multiple fields at once.

In this case, we are not breaking the rule of rejecting with an error, there just happens to be more than one 😉 .

I understand that this may be an edge case that doesn't fit into the "work for 90% of cases 100% of the time" rule. I just wanted to flag up this use case as one of perhaps a set of cases where you might want to be able to suppress the warnings only sometimes, and put in a request for this feature.

We're currently getting lots of warnings, but I'd love to have some mechanism whereby we could only get them when we're not returning an array but for example inadvertently returning something like `undefined` as that is super useful. 

Thanks for listening, we ❤️  Bluebird.

**Comment by benjamingr** [comment]
@EricDS what about rejecting with an AggregateError (an error object that is also an array like of errors)?

**Comment by petkaantonov** [comment]
Yep there's [Promise.AggregateError](http://bluebirdjs.com/docs/api/aggregateerror.html) that implements array and error interfaces

**Comment by ben-page** [comment]
So the last few message are off topic. This issue is about providing a way for the author of method for flagging their promises as handled. Rather than requiring the users of the method to suppress warnings with `return null`. Currently, a runaway promise warning can be triggered when code that uses Bluebird consumes modules that also uses Bluebird and the module author has no control over that.


```javascript
function fakeWork() {
    return new Promise(resolve => {
        resolve();
    });
}

//module 1
function doSomething1() {
    fakeWork()
        .then(() => {
            console.log('did thing');
        })
        .catch(err => {
            //totally handle error
        })
        .suppressRunawayWarning(); //proposed way to suppress the warning as the producer
}

//module 2
function doSomething2() {
    return fakeWork()
        .then(() => {
            doSomething1(); //calling another module, not away that it uses Bluebird

            //return null; //user of module is required to understand it's design and suppress warning
        });
}

doSomething2()
    .then(() => {
        console.log('did something 2');
    });
```

**Comment by ben-page** [comment]
@spion, `done()` does not solve this case, the warning is still raised even if done is called where I have `suppressRunawayWarning`.

**Comment by ErisDS** [comment]
> I'd also like to ask about an additional way to suppress or configure the warnings thrown from Bluebird, but with regard to the a promise was rejected with a non-error warning.

This was not off-topic IMO. The particular use-case I presented was provided with an alternate answer or dismissed depending on how you look at it ;) However, I believe there needs to be a way to suppress individual cases of all the warnings that Bluebird can throw and the solution should be considered as a single thing.

**Comment by spion** [comment]
@ben-page the thing is, `done()` should actually supress the warning. Once you are done with a promise, it should not matter where you created it or whether you returned it or not. So I don't think it should be necessary to add a new method for this.

**Comment by baffo32** [comment]
I ran into this.  I wanted to instrument `setTimeout` etc to use promises, so as to get bluebird's expanded stack traces, but the only way to do it without triggering these warnings is to much about inside bluebird's internals, because these functions do not return promises.  As stated earlier, `done()` is a great place to put this solution, but it does not work as-is yet.

**Comment by fgarcia** [comment]
IMHO it should be possible to disable warnings for third party code.

Right now I depend on a library to test each of my React components and my test reports are full of warnings about runways promises. It is just a one line fix returning null instead undefined, but the project maintainer closed the issue arguing it is a bug in Blueprint and suggested to stop using Bluebird and switch to native Promises

**Comment by petkaantonov** [comment]
Warnings can be disabled with environment variables or .config

```js
//module 2
function doSomething2() {
    return fakeWork()
        .then(() => {
            doSomething1(); //calling another module, but not aware that it uses Bluebird internally

            //return null; //user of module is required to understand it's internals and suppress warning
        });
}
````
doSomething returns a promise and you are inside then, seems like the warning is valid. If your chain ends here you could use .done.
